classes/TaskRunner/Upgrade/UpgradeFiles.php

```diff
@@ -0,0 +1,284 @@
+        }
+        
+        // list saved in UpgradeFileNames::toUpgradeFileList
+        // get files differences (previously generated)
+        $admin_dir = trim(str_replace($this->container->getProperty(UpgradeContainer::PS_ROOT_PATH), '', $this->container->getProperty(UpgradeContainer::PS_ADMIN_PATH)), DIRECTORY_SEPARATOR);
+        $filepath_list_diff = $this->container->getProperty(UpgradeContainer::WORKSPACE_PATH).DIRECTORY_SEPARATOR.UpgradeFileNames::diffFileList;
+        $list_files_diff = array();
+        if (file_exists($filepath_list_diff)) {
+            $list_files_diff = $this->container->getFileConfigurationStorage()->load(UpgradeFileNames::diffFileList);
+            // only keep list of files to delete. The modified files will be listed with _listFilesToUpgrade
+            $list_files_diff = $list_files_diff['deleted'];
+            foreach ($list_files_diff as $k => $path) {
+                if (preg_match("#autoupgrade#", $path)) {
+                    unset($list_files_diff[$k]);
+                } else {
+                    $list_files_diff[$k] = str_replace('/'.'admin', '/'.$admin_dir, $path);
+                }
+            } // do not replace by DIRECTORY_SEPARATOR
+        }
+
+        if (!($list_files_to_upgrade = $this->listFilesToUpgrade($this->container->getProperty(UpgradeContainer::LATEST_PATH)))) {
+            return false;
+        }
+
+        // also add files to remove
+        $list_files_to_upgrade = array_merge($list_files_diff, $list_files_to_upgrade);
+
+        $filesToMoveToTheBeginning = array(
+            DIRECTORY_SEPARATOR.'vendor'.DIRECTORY_SEPARATOR.'autoload.php',
+            DIRECTORY_SEPARATOR.'vendor'.DIRECTORY_SEPARATOR.'composer'.DIRECTORY_SEPARATOR.'ClassLoader.php',
+            DIRECTORY_SEPARATOR.'vendor'.DIRECTORY_SEPARATOR.'composer'.DIRECTORY_SEPARATOR.'autoload_classmap.php',
+            DIRECTORY_SEPARATOR.'vendor'.DIRECTORY_SEPARATOR.'composer'.DIRECTORY_SEPARATOR.'autoload_files.php',
+            DIRECTORY_SEPARATOR.'vendor'.DIRECTORY_SEPARATOR.'composer'.DIRECTORY_SEPARATOR.'autoload_namespaces.php',
+            DIRECTORY_SEPARATOR.'vendor'.DIRECTORY_SEPARATOR.'composer'.DIRECTORY_SEPARATOR.'autoload_psr4.php',
+            DIRECTORY_SEPARATOR.'vendor'.DIRECTORY_SEPARATOR.'composer'.DIRECTORY_SEPARATOR.'autoload_real.php',
+            DIRECTORY_SEPARATOR.'vendor'.DIRECTORY_SEPARATOR.'composer'.DIRECTORY_SEPARATOR.'autoload_static.php',
+            DIRECTORY_SEPARATOR.'vendor'.DIRECTORY_SEPARATOR.'composer'.DIRECTORY_SEPARATOR.'include_paths.php',
+            DIRECTORY_SEPARATOR.'vendor'.DIRECTORY_SEPARATOR.'composer',
+            DIRECTORY_SEPARATOR.'vendor',
+        );
+
+        foreach ($filesToMoveToTheBeginning as $file) {
+            if ($key = array_search($file, $list_files_to_upgrade)) {
+                unset($list_files_to_upgrade[$key]);
+                $list_files_to_upgrade = array_merge(array($file), $list_files_to_upgrade);
```

Comment:
Instead of using array_merge, you can use `$listFilesToUpgrade[] = $file;`

Reply:
No, read the previous line and the name of the different variables.
We absolutely need to move these files at the *beginning* of the array.